File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/request/TaskBuilder.java
package com.roxiemobile.networkingapi.network.rest.request;

import com.roxiemobile.networkingapi.network.rest.Task;

public interface TaskBuilder<Ti, To> {

    /**
     * TODO
     */
    String tag();

    /**
     * The original HTTP request.
     */
    RequestEntity<Ti> requestEntity();

    /**
     * TODO
     */
    Task<Ti, To> build();
}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/request/TaskBuilder.kt
package com.roxiemobile.networkingapi.network.rest.request

import com.roxiemobile.networkingapi.network.http.body.HttpBody
import com.roxiemobile.networkingapi.network.rest.Task
import com.roxiemobile.networkingapi.network.rest.configuration.HttpClientConfig

interface TaskBuilder<Ti: HttpBody, To>: Cloneable {

// MARK: - Methods

    /**
     * TODO
     */
    fun tag(tag: String?): TaskBuilder<Ti, To>

    /**
     * TODO
     */
    fun requestEntity(requestEntity: RequestEntity<Ti>): TaskBuilder<Ti, To>

    /**
     * TODO
     */
    fun httpClientConfig(httpClientConfig: HttpClientConfig): TaskBuilder<Ti, To>

    /**
     * TODO
     */
    fun build(): Task<Ti, To>

    /**
     * TODO
     */
    public override fun clone(): TaskBuilder<Ti, To>
}
